llvm/lib/SYCLLowerIR/TargetHelpers.cpp

```diff
@@ -82,6 +86,36 @@ void populateKernels(Module &M, SmallVectorImpl<KernelPayload> &Kernels,
       if (auto *Func = dyn_cast<Function>(FuncConstant->getValue()))
         Kernels.push_back(KernelPayload(Func, MetadataNode));
   }
+
+  // We need to match non-kernel metadata nodes using the kernel name to the
+  // kernel nodes.
+  for (auto &KP : Kernels) {
+    auto *KernelConstant = dyn_cast<ConstantAsMetadata>(KP.MD->getOperand(0));
+    assert(KernelConstant);
+    auto KernelName =
+        dyn_cast<Function>(KernelConstant->getValue())->getFunction().getName();
+    // Keep track of matched nodes, to decrease the search space with each
+    // iteration.
+    SmallVector<unsigned> ToDelete;
+    for (unsigned i = 0; i < PossibleDependencies.size(); ++i) {
+      auto *Dep = PossibleDependencies[i];
+      const MDOperand &FuncOperand = Dep->getOperand(0);
+      if (!FuncOperand)
+        continue;
+      if (auto *FuncConstant = dyn_cast<ConstantAsMetadata>(FuncOperand))
+        if (auto *Func = dyn_cast<Function>(FuncConstant->getValue()))
+          // We've found a match, append the dependent node to the kernel
+          // payload.
+          if (KernelName == Func->getFunction().getName()) {
+            KP.DependentMDs.push_back(Dep);
+            ToDelete.push_back(i);
+          }
+
+      // Remove matched dependencies.
+      for (auto i = ToDelete.rbegin(); i != ToDelete.rend(); ++i)
```

Comment:
This `i` is shadowing variable `i` from the outer loop.
Additionally, deleting from a vector is rather a slow operation, but I guess we can't use `set` here, because we won't be able to iterate and delete from it at the same time.

Should we use `ToDelete` as a mask and simply ignore indices `i`, which are within `ToDelete`? It could either be resized to match `PossibleDependencies` or it could be converted into a set.

Reply:
Agreed, replaced with a set approach. Great find with the shadowing, kinda hoped compiler would warn me about that...